### ISSUE TITLE
Balance change does not trigger payout for indirectly staked account

### ISSUE THREAD

**Issue by robkainz** [hashgraph/hedera-services#4166]
### Description

If account A stakes to account B and account B stakes to node N, according to Hedera documentation "When your staked account has hbars transferred to it or debited from it" or “When the total number of hbars staked to an account changes”, the payout should be triggered. However, crediting or debiting account A's balance does not trigger the rewards payment due account B. I tried this several times in production and it fails, yet if the payment is made to the account which is directly staked, then the payout happens. I realize there may be different interpretations of the above statements from Hedera's documentation and possibly this is "working as designed", if that's the case we would like clarification before we start charging people for collecting rewards on accounts involving indirect staking.

### Steps to reproduce

1. Create new account B and stake it to node 1.
2. Create new account A and stake it to account B.
3. Send payment from some other account C to account A. (Rewards payout is not triggered as I would expect from my understanding of Hedera's reward payout triggers.)

### Hedera network

mainnet

### Version

mainnet production version

### Operating system

macOS

**Comment by Neeharika-Sompalli** [comment]
As per the HIP, these are the conditions rewards are paid.

> At the start of a given staking period (for example, at midnight UTC at the start of a one-day period), a snapshot is taken to determine the total amount staked to each node at that moment. This determines the weight of each node in consensus throughout the duration of that staking period. At the end of the staking period, all stakers who should earn a reward will become eligible to receive that earned reward. The earned reward will actually be transferred the next time the staking account changes, such as when it pays for a transaction, or has hbars transferred to it, or there are hbars transferred to an account that stakes to it, or it auto-renews because it reached its expiration date (which happens at least every 3 months according to current settings). With current settings, this means it will receive rewards at most once per day, and least once every 3 months.

From the scenario, it seems like account B is not staked to node at start of staking period. At the end of first staking period (24 hours currently), the account becomes eligible for reward. 
If there is a change in account A's balance in the next staking period, rewards will be paid. Also, please note that within the same period rewards are not awarded twice.

**Comment by lbaird** [comment]
Or, to put it in other words: do steps 1 and 2 on a Monday and step 3 on
Wednesday, and it will trigger a reward. Unless you already triggered the
Wednesday reward. Each day is the period from a midnight UTC to the next
midnight UTC.

**Comment by robkainz** [comment]
@lbaird @Neeharika-Sompalli I'm sorry, but I configured my indirectly staked account months ago, when Wallawallet first supported staking (first day staking config was supported). I tried this again sending 1 tinybar from unrelated account C to account B and it did not trigger any payout to account A as expected, even though my account A shows pending rewards > 0 (via the direct Hedera network query, but this is another issue. I believe the mirror node is INCORRECT as it's returning 0 hbar pending rewards and doesn't match. Also, the start date is a day earlier on the paid Hedera account info query than on the mirror, yet I believe the mirror date is actually correct.) I haven't opened the other defects, because I don't want to share my account information, but will do so privately with you if you want them. Greg Scullard has this info already via Discord but he's on vacation (though he's aware.)

**Comment by lbaird** [comment]
If you’re sure that A didn’t get a reward triggered earlier that day, then
it would be a bug on mainnet for the transfer to B to fail to trigger the
reward to A. It would be good to verify that nothing else triggered a
reward to A, earlier that day.

**Comment by robkainz** [comment]
@lbaird I am 100% certain nothing triggered a reward earlier that day. I am connected to you on Linked In (for a long time), and I know you don't read those messages anymore, but I'm happy to send you my account id there if you want to investigate further.

**Comment by robkainz** [comment]
@lbaird Here are the important steps taken:

On 10/31 just after UTC midnight I sent 1 tinybar from A to B and a reward was paid to B (as expected.)

On morning of November 2 (after 8 AM ET - aka after UTC midnight), the mirror shows a start date for 10/31 and Hedera services (paid query) shows a start date of 10/30. I believe this is a defect and the Hedera services should return 10/31.) What's peculiar is that the Hedera services query showed a pending reward (and still does) whereas the mirror showed 0 hbar pending reward (and still does).

On November 2 (after 8 AM ET, or after UTC midnight) I sent 1 tinybar from C to A. No reward payout was triggered for B, but should have been (as far as I understand). My understanding is the staking period should have started October 31 midnight UTC (as I believe the mirror correctly shows, but Hedera services does not) and ended November 1 midnight UTC so there should have been rewards available on November 2. 

I've been connected on LinkedIn to you for years, and I know you don't check your messages there, but I will send you my A, B, C account ids so you can investigate further if you ask me to.

**Comment by robkainz** [comment]
@lbaird Just to clarify, the last time 0.0.800 shows up in the transfers for a transaction for A was October 31, just after UTC midnight. As I said, you can get the account numbers from me, or Greg Scullard.

**Comment by tinker-michaelj** [comment]
> @lbaird Just to clarify, the last time 0.0.800 shows up in the transfers for a transaction for A was October 31, just after UTC midnight. As I said, you can get the account numbers from me, or Greg Scullard.

Hi @robkainz, can you double-check all of B's transactions in this time period?

One other comment; you wrote,
> On morning of November 2 (after 8 AM ET), the mirror shows a start date for 10/31 and Hedera services (paid query) shows a start date of 10/30. (I believe this is a defect and the Hedera services should return 10/31.)

10/30 is the correct updated "start period start" in this scenario, because B remains eligible for rewards on 10/31. (This is a bit counter-intuitive, but is the expected behavior.)

**Comment by tinker-michaelj** [comment]
> @lbaird @Neeharika-Sompalli I'm sorry, but I configured my indirectly staked account months ago, when Wallawallet first supported staking (within first days staking config was supported).

It would also be good to validate A's staking election via a paid query, since we've observed cases of a client supporting transactions using new protobuf fields **before the consensus nodes were upgraded to also support those fields**.

**Comment by lbaird** [comment]
Yes, if you can send me the 3 account numbers, and it’s ok for me to share
it with the engineers, then they can figure out what’s going on.

**Comment by robkainz** [comment]
> Yes, if you can send me the 3 account numbers, and it’s ok for me to share it with the engineers, then they can figure out what’s going on.
> […](#)

@lbaird I did send you the account ids via linkedin.

**Comment by robkainz** [comment]
> > @lbaird Just to clarify, the last time 0.0.800 shows up in the transfers for a transaction for A was October 31, just after UTC midnight. As I said, you can get the account numbers from me, or Greg Scullard.
> 
> Hi @robkainz, can you double-check all of B's transactions in this time period?
> 
> One other comment; you wrote,
> 
> > On morning of November 2 (after 8 AM ET), the mirror shows a start date for 10/31 and Hedera services (paid query) shows a start date of 10/30. (I believe this is a defect and the Hedera services should return 10/31.)
> 
> 10/30 is the correct updated "start period start" in this scenario, because B remains eligible for rewards on 10/31. (This is a bit counter-intuitive, but is the expected behavior.)

Please ask Dr. Baird or Greg Scullard for my account numbers, or reach out to me directly as needed so you can see everything. After the rewards payout on October 31, just after midnight UTC, the only other transaction is November 1 in the afternoon EDT on November 1, and it's a small transfer from account A to B, resulting in no rewards payout. This was an attempt to trigger the payout, but I suspected I may need to wait another day, because the payout already occurred just after midnight UTC on October 31 (on November 1 effectively) so I may need to wait another 24 hour period after the next midnight UTC, which would be tonight at 8 PM my time. Is that correct? If that's the case though, should I see any pending rewards at all from the paid query? I am confused, because on the morning of November 2 the paid query showed me the expected number of hbar rewards, but the mirror node showed me 0. Now I'm wondering if the mirror node is right there too, since I couldn't trigger the reward yesterday. I have not attempted to trigger the reward today yet, so that I could keep things in this state in case you need to look at it for diagnostic purposes. I suspect though that if I send a small payment directly to account B that the payout may happen. (Please feel free to send some amounts to that account in an attempt to trigger the payments, preferably enough so I can pay my cohort developer another day's wages while we work through this confusion caused by the mirror/Hedera mismatches!)

10/30 is definitely counter-intuitive and in any case that paid query doesn't match the mirror - so I guess that means the mirror is wrong. So who is right about the pending rewards then? I guess another trigger payment sent to account B before 8 PM EDT today will help us determine that.

**Comment by robkainz** [comment]
@tinker-michaelj 
> > @lbaird @Neeharika-Sompalli I'm sorry, but I configured my indirectly staked account months ago, when Wallawallet first supported staking (within first days staking config was supported).
> 
> It would also be good to validate A's staking election via a paid query, since we've observed cases of a client supporting transactions using new protobuf fields **before the consensus nodes were upgraded to also support those fields**.

Sure, I queried account A's account info using Hedera SDK golang version v2.17.6, the version currently used by Wallawallet in production. I retrieved the account information for account A using the paid query and verified that it is not staked to a node id, but is staked to account B's account id, as expected. Other fields that show up for that paid query are: Staked to Me: 0 hbar, Stake Period Start: December 31, 0001 07:03:58 PM GMT-04:56 (obviously an invalid timestamp but I don't believe this is a bug in Wallawallet - and it will not be shown for an account staked to another account in a future version), Pending Reward 0 hbar and Decline Staking Reward false. If you need any other info from this account, I will need to step through the code and grab the results of the account info query as it comes back from the node directly.

**Comment by tinker-michaelj** [comment]
Hey @robkainz, appreciate this information!

There's a known defect with mirror node's pending rewards now, so we should ignore them for the purposes of this investigation.

>  I retrieved the account information for account A using the paid query and verified that it is not staked to a node id, but is staked to account B's account id, as expected.

Got it...what do you see when you do the same query for account B now?

**Comment by robkainz** [comment]
@tinker-michaelj 
> Hey @robkainz, appreciate this information!
> 
> There's a known defect with mirror node's pending rewards atm, so we should ignore them for the purposes of this investigation.
> 
> > I retrieved the account information for account A using the paid query and verified that it is not staked to a node id, but is staked to account B's account id, as expected.
> 
> Got it...what do you see when you do the same query for account B now?

You're welcome! Thanks for looking into this and if you don't trigger the rewards payment by tonight at 8 PM EST, I will do it somewhere between 7:30 PM and before 8 PM.

Node B gives me: Staked Node Id: 1, Staked Account Id: (none), Staked to Me: xxxx hbar (>0), Stake Period Start: October, 30 2022 08:00:00 PM GMT-04:00, Pending Reward: yyyy hbar (>0), Decline Staking Reward: false

**Comment by tinker-michaelj** [comment]
@robkainz That's very interesting. So we'd expect that if you send e.g. 1 hbar to account A (even right now), then two things should happen:
  1. B's `stakedToMe` should increase by 1 hbar.
  2. B should receive the rewards you currently see as pending.
  3. B's `stakePeriodStart` should revert to UTC November 1.

**Comment by robkainz** [comment]
> @robkainz That's very interesting. So we'd expect that if you send e.g. 1 hbar to account A (even right now), then two things should happen:
> 
> 1. B's `stakedToMe` should increase by 1 hbar.
> 2. B should receive the rewards you currently see as pending.
> 3. B's `stakePeriodStart` should revert to UTC November 1.

Yes, I think all 3 of those things should happen if I understand the mechanism correctly (if I designed it for indirect staking, that is what I would expect). When you are ready to test it, please send some hbar to account A, or let me know and I will do it.

**Comment by tinker-michaelj** [comment]
Hey @robkainz I haven't got the A/B/C id's from Greg yet so feel free!

**Comment by robkainz** [comment]
> Hey @robkainz I haven't got the A/B/C id's from Greg yet so feel free!

All this time I was assuming a tinybar transfer would be all I need to do, so I sent 1 tinybar to my A account. This had no effect, and the B "Staked to Me" only shows HBAR, not tinybar (basically I see ###.00000000 - all zeros for fractional). So, I'm guessing the change in balance is not detected for fractions, only whole hbars. Is this by design? Perhaps sending a whole hbar will work as you expect, but personally I would expect a single tinybar would also make the difference. I'll send a whole hbar next and report...

**Comment by lbaird** [comment]
All staked values are truncated down to the nearest integer hbar. So adding
a tinybar only affects staked amount if it pushes the balance over an
integer boundary. Sending 1 hbar will certainly push it over a boundary.

**Comment by tinker-michaelj** [comment]
@robkainz Thanks for running the experiment! As @lbaird noted, all staked amounts are rounded to the nearest hbar (see the HIP-406 section [here](https://hips.hedera.com/hip/hip-406#19-accounts-stake-non-whole-number-of-hbars)).

So if `A` stakes to `B`, and `A` receives just one tinybar, in general `B`'s `stakedToMe` amount won't change.

And in the current code, `B`'s `stakedToMe` amount **does have to change** in order for it to receive rewards in this situation.

But I see your point that this is confusing. Arguably we should change so that whenever A's balance changes---even by a tinybar---B will receive rewards.

How do you feel, having been through this journey? 😁

**Comment by robkainz** [comment]
@tinker-michaelj @lbaird @Neeharika-Sompalli 

It’s been fun everyone, lol. Thanks for being patient with me, it’s been a while since I read the spec and obviously this point didn’t sink in. 

The bigger problem for us is the mirror node inconsistencies, because we can alway determine the correct node either directly or indirectly to send 1 tinybar to to trigger the payout, but our users are confused about when to expect those payouts, and without correctly returning the pending rewards and the correct start date from the mirror it is difficult to provide them with a countdown timer so they know when to expect the reward without first charging them for the paid account info query (which depending on configuration of their payment account can actually trigger the payout when they didn’t realize it, then resetting pending rewards to zero, or maybe that 0 is the known mirror bug.) We would rather make a small fee on the collect rewards button and simplify the user experience with the countdown timer. Due to this confusion, some users have reconfigured their staking multiple times, paying our larger fee when not necessary. We definitely want to help them avoid that.

Thanks for all you guys are doing!

**Comment by tinker-michaelj** [comment]
@robkainz Absolutely, appreciate your time as well. 

I know mirror team is working hard on this; I'll update you here as soon as I have a timeline on the fix

**Comment by robkainz** [comment]
@tinker-michaelj @lbaird One thing I also noticed is that the mirror node didn’t have the “Staked to Me” hbar data, do you know if there  are plans to add it too?

**Comment by tinker-michaelj** [comment]
Hi @robkainz, I have two updates:
1. You can follow [this mirror node ticket](https://github.com/hashgraph/hedera-mirror-node/issues/4815) for details and progress on the pending rewards issue.
2. The same performance challenges that currently force mirror node to update pending rewards only once-per-UTC-day would apply to `stakedToMe` calculations. Until these performance issues are resolved (which could be some time), mirror node does not plan to publish `stakedToMe` information.
